twill-core/src/main/java/org/apache/twill/internal/json/ResourceReportAdapter.java

```diff
@@ -59,4 +72,61 @@ public ResourceReport fromJson(String json) {
   public ResourceReport fromJson(Reader reader) {
     return gson.fromJson(reader, ResourceReport.class);
   }
+
+  /**
+   * A type adapter for serializing {@code Collection<TwillRunResources>} correctly
+   */
+  private static final class ResourceReportTypeAdapterFactory implements TypeAdapterFactory {
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public <T> TypeAdapter<T> create(Gson gson, TypeToken<T> type) {
+      if (!Collection.class.isAssignableFrom(type.getRawType())) {
+        return null;
+      }
+
+      if (!(type.getType() instanceof ParameterizedType)) {
+        return null;
+      }
+
+      TypeToken<?> valueType = TypeToken.get(((ParameterizedType) type.getType()).getActualTypeArguments()[0]);
+      return (TypeAdapter<T>) collectionAdapter(gson, valueType);
+    }
+
+    private <V> TypeAdapter<Collection<V>> collectionAdapter(Gson gson, TypeToken<V> valueType) {
+      final TypeAdapter<V> valueAdapter = gson.getAdapter(valueType);
+      return new TypeAdapter<Collection<V>>() {
+        @Override
+        public void write(JsonWriter writer, @Nullable Collection<V> collection) throws IOException {
+          if (collection == null) {
+            writer.nullValue();
+            return;
+          }
+          writer.beginArray();
+          for (V value : collection) {
+            valueAdapter.write(writer, value);
+          }
+          writer.endArray();
+        }
+
+        @Override
+        public Collection<V> read(JsonReader reader) throws IOException {
+          if (reader.peek() == JsonToken.NULL) {
+            reader.nextNull();
+            return null;
+          }
+          if (reader.peek() != JsonToken.BEGIN_ARRAY) {
+            return null;
```

Comment:
Same as above, we only need to generate result when we encounter `Collection<TwillRunResources>`